packages/gensx-core/src/restore-checkpoint.ts

```diff
@@ -0,0 +1,95 @@
+/* eslint-disable @typescript-eslint/no-unnecessary-type-parameters */
+import { Component } from "./component.js";
+import { getCurrentContext } from "./context.js";
+
+// This is a magical component used to store a marker in the checkpoint. When restoring from a checkpoint, the execution layer will update this marker in the checkpoint with the given feedback,
+// remove all subsequent nodes, and resume execution from this point.
+const CheckpointMarkerComponent = Component(
+  `CheckpointMarker`,
+  ({ maxRestores }: { maxRestores: number }) => {
+    const context = getCurrentContext();
+    const currentNodeId = context.getCurrentNodeId();
+    if (!currentNodeId) {
+      throw new Error(`[GenSX] No current node id found.`);
+    }
+
+    return {
+      feedback: null,
+      restoreCount: 0,
+      maxRestores,
+      nodeId: currentNodeId,
+    };
+  },
+);
+
+export function createCheckpoint<T = unknown>(
+  {
+    label,
+    // schema,
+  }: {
+    label?: string;
+    // schema: z.ZodSchema<T> // TODO: Validate the given feedback against the schema
+  } = {},
+  { maxRestores = 3 }: { maxRestores?: number } = {},
+): {
+  feedback: T | null;
+  restore: (feedback: T) => Promise<void>;
+  label: string;
+} {
+  label ??= `checkpoint-marker-${Date.now()}`;
+  const context = getCurrentContext();
+  const workflowContext = context.getWorkflowContext();
+  if (workflowContext.checkpointLabelMap.has(label)) {
+    throw new Error(`[GenSX] Checkpoint ${label} has already been created.`);
+  }
+  // Do not pass the label as a prop, as that would affect that ability to deterministically calculate the nodeId.
+  const result = CheckpointMarkerComponent(
+    { maxRestores },
+    {
+      metadata: { label, maxRestores },
+    },
+  );
+  workflowContext.checkpointLabelMap.set(label, result.nodeId);
+
+  if (result.restoreCount >= result.maxRestores) {
+    throw new Error(
+      `[GenSX] Checkpoint ${label} has been restored more than ${result.maxRestores} times.`,
+    );
+  }
+
+  return {
+    feedback: result.feedback as T | null,
+    restore: (feedback: T) =>
```

Comment:
Is it safe to pass this `restore` function around in the workflow? IE can I pass it to another component? 

If the program crashes and replays will this function still be "intact"? IE `json.parse` on the outputs of this component actually rehydrates the restore function...